src/Middleware/RequestQueue/test/SemaphoreWrapperTests.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using System.Threading;
+using System.Threading.Tasks;
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Testing;
+
+namespace Microsoft.AspNetCore.RequestQueue.Tests
+{
+    public class SemaphoreWrapperTests
+    {
+        [Fact]
+        public async Task TestBehavior()
+        {
+            var s = new SemaphoreWrapper(1);
+            Assert.Equal(1, s.SpotsLeft);
+
+            await s.EnterQueue();
+            Assert.Equal(0, s.SpotsLeft);
+
+            s.LeaveQueue();
+            Assert.Equal(1, s.SpotsLeft);
+        }
+
+        [Fact]
+        public void TestQueueLength()
```

Comment:
As @jkotalik mentioned, we like our test names to be nice and descriptive. When the test fails we want someone who sees only the test name to have a relatively good idea of what's going on.

For example here it would be: "EnterQueueBlocksWhenNoQueueSpaceIsAvailable"

Don't be afraid to have long method names when it comes to tests, within some reason. For example "ThisTestEnsuresThatIfYouCallEnterQueueAndThereAreNoSpotsTheTaskReturnedIsNotCompleted" would be **too** verbose 😝